### DISCUSSION TITLE
Migration from mplex to yamux

### DISCUSSION THREAD

**Discussion by odesenfans** [libp2p/rust-libp2p#3916] [Q&A] [answered]
Hello,

I just upgraded my P2P stack to Rust libp2p v0.51.3. One of the big changes in that release is the deprecation of `mplex` in favor of `yamux`, which seems to make sense. However, I am now confronted with the difficulty of upgrading the P2P network from mplex to yamux smoothly, meaning that new peers running with v0.51.3 should be able to dial up-to-date peers using yamux and outdated peers using mplex.

What would be the best way to achieve this? Is this possible at all?

**Comment by thomaseizinger** [comment, marked as answer]
You can use `SelectUpgrade` where you list `yamux` first, meaning it will be the preferred muxer.

We will continue to maintain `libp2p-mplex` but you'll have to depend on it yourself, the reexport from `libp2p` will be gone in the next release :)

**Comment by dariusc93** [comment]
Pretty much like @thomaseizinger said, you could use `SelectUpgrade`, with `yamux` listed first. If this stack is configurable by the end user, you could use `Either` to use just `yamux` or both muxers. Eg https://github.com/dariusc93/rust-ipfs/blob/libp2p-next/src/p2p/transport.rs#L172-L177 .

**Comment by odesenfans** [comment]
Great, I figured the mechanism was there somewhere but I couldn't find it. It seems to work! Thanks for the example @dariusc93.
